File: src/services/llm_client.py
"""Client for LLM (Large Language Model) service"""
import sys
import asyncio
import aiohttp
from typing import Optional, List, Dict


class LLMClient:
    """Single-request LLM client with streaming sentence support"""

    def __init__(self, host: str, port: int, model: str, temperature: float, max_tokens: int):
        self.base_url = f"http://{host}:{port}"
        self.model = model
        self.temperature = temperature
        self.max_tokens = max_tokens
        self.session: Optional[aiohttp.ClientSession] = None

        # Current request state
        self.current_request: Optional[List[Dict[str, str]]] = None
        self.pending_sentences: List[str] = []

        # Control flags
        self.is_generating = False
        self.should_abort = False
        self._processing_task: Optional[asyncio.Task] = None
        self._started = False

    async def __aenter__(self):
        self.session = aiohttp.ClientSession()
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        if self.session:
            await self.session.close()

    def start(self) -> None:
        """Start the processing loop (call once inside async context)"""
        if not self._started:
            self._processing_task = asyncio.create_task(self._processing_loop())
            self._started = True

    def enqueue(self, messages: List[Dict[str, str]]) -> None:
        """
        Set new LLM request (replaces any in-progress).
        Non-blocking.
        """
        if not self._started:
            raise RuntimeError("LLMClient not started. Call start() first in async context.")

        # Abort any current request
        if self.current_request is not None:
            self.should_abort = True

        # Set new request
        self.current_request = messages
        self.should_abort = False

    def get_ready_sentences(self) -> List[str]:
        """
        Get all completed sentences and clear the list (non-blocking).
        Returns list of sentence strings.
        """
        if not self.pending_sentences:
            return []

        result = self.pending_sentences.copy()
        self.pending_sentences.clear()
        return result

    def is_processing(self) -> bool:
        """Returns True if currently generating or has pending sentences"""
        return self.is_generating or len(self.pending_sentences) > 0 or self.current_request is not None

    def abort(self) -> None:
        """Cancel current request and clear pending sentences (non-blocking)"""
        self.current_request = None
        self.pending_sentences.clear()
        self.should_abort = True

    async def _processing_loop(self):
        """Background task that processes LLM requests"""
        while True:
            try:
                # Wait for a request
                if self.current_request is None:
                    await asyncio.sleep(0.01)
                    continue

                # Process the request
                messages = self.current_request
                self.current_request = None  # Clear immediately so new requests can come in

                await self._generate(messages)

            except asyncio.CancelledError:
                break
            except Exception as e:
                print(f"LLM processing loop error: {e}", file=sys.stderr)
                import traceback
                traceback.print_exc()

    async def _generate(self, messages: List[Dict[str, str]]):
        """Generate LLM response with streaming"""
        if not self.session:
            raise RuntimeError("LLMClient not initialized. Use 'async with' context manager.")

        self.is_generating = True
        self.should_abort = False

        payload = {
            "model": self.model,
            "messages": messages,
            "temperature": self.temperature,
            "max_tokens": self.max_tokens,
            "stream": True  # Enable streaming
        }

        try:
            async with self.session.post(
                f"{self.base_url}/v1/chat/completions",
                json=payload,
                timeout=aiohttp.ClientTimeout(total=120)
            ) as response:
                response.raise_for_status()

                # Stream the response
                accumulated_text = ""
                async for line in response.content:
                    if self.should_abort:
                        break

                    # Parse SSE format
                    line = line.decode('utf-8').strip()
                    if not line or not line.startswith('data: '):
                        continue

                    data_str = line[6:]  # Remove 'data: ' prefix
                    if data_str == '[DONE]':
                        break

                    try:
                        import json
                        data = json.loads(data_str)

                        # Extract content delta
                        if 'choices' in data and len(data['choices']) > 0:
                            delta = data['choices'][0].get('delta', {})
                            content = delta.get('content', '')

                            if content:
                                accumulated_text += content

                                # Check for sentence boundaries
                                sentences = self._extract_sentences(accumulated_text)
                                if sentences:
                                    for sentence in sentences[:-1]:  # All but last (might be incomplete)
                                        if sentence.strip():
                                            self.pending_sentences.append(sentence.strip())
                                    accumulated_text = sentences[-1]  # Keep incomplete part

                    except json.JSONDecodeError:
                        continue

                # Add any remaining text as final sentence
                if accumulated_text.strip() and not self.should_abort:
                    self.pending_sentences.append(accumulated_text.strip())

        except asyncio.CancelledError:
            pass
        except Exception as e:
            print(f"LLM API error: {e}", file=sys.stderr)
            import traceback
            traceback.print_exc()
        finally:
            self.is_generating = False

    def _extract_sentences(self, text: str) -> List[str]:
        """
        Split text on sentence boundaries.
        Returns list where last item might be incomplete.
        """
        # Simple sentence splitting on .!?
        import re
        # Split but keep delimiters
        parts = re.split(r'([.!?]+)', text)

        sentences = []
        for i in range(0, len(parts) - 1, 2):
            sentence = parts[i]
            punct = parts[i + 1] if i + 1 < len(parts) else ""
            sentences.append(sentence + punct)

        # Add remaining text (incomplete sentence)
        if len(parts) % 2 == 1:
            sentences.append(parts[-1])
        elif not sentences:
            sentences.append(text)

        return sentences

    async def stop(self):
        """Stop the processing loop (cleanup)"""
        if self._processing_task:
            self._processing_task.cancel()
            try:
                await self._processing_task
            except asyncio.CancelledError:
                pass

File: rust-earshot/src/clients/llm_client.rs
//! HTTP/SSE client for LLM API (OpenAI-compatible)
//!
//! Protocol:
//! - POST /v1/chat/completions with JSON body
//! - Response is SSE stream: data: {...}
//! - Parse choices[0].delta.content for text chunks
//! - data: [DONE] marks end of stream

use anyhow::{Context, Result};
use futures_util::StreamExt;
use reqwest::Client;
use serde::{Deserialize, Serialize};
use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
use std::sync::Arc;
use tokio::sync::Mutex;
use tracing::{debug, error, info, warn};

use crate::config::LlmConfig;

/// Chat message
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChatMessage {
    pub role: String,
    pub content: String,
}

impl ChatMessage {
    pub fn system(content: &str) -> Self {
        Self {
            role: "system".to_string(),
            content: content.to_string(),
        }
    }

    pub fn user(content: &str) -> Self {
        Self {
            role: "user".to_string(),
            content: content.to_string(),
        }
    }

    pub fn assistant(content: &str) -> Self {
        Self {
            role: "assistant".to_string(),
            content: content.to_string(),
        }
    }
}

/// Request body for chat completions
#[derive(Debug, Serialize)]
struct ChatCompletionRequest {
    model: String,
    messages: Vec<ChatMessage>,
    temperature: f32,
    max_tokens: i32,
    stream: bool,
}

/// Delta content from streaming response
#[derive(Debug, Deserialize)]
struct Delta {
    content: Option<String>,
}

/// Choice from streaming response
#[derive(Debug, Deserialize)]
struct Choice {
    delta: Delta,
}

/// Streaming response chunk
#[derive(Debug, Deserialize)]
struct StreamChunk {
    choices: Vec<Choice>,
}

/// LLM client with streaming and cancellation support
pub struct LlmClient {
    url: String,
    model: String,
    temperature: f32,
    system_prompt: String,
    client: Client,

    // Pending sentences (complete sentences ready for TTS)
    pending_sentences: Arc<Mutex<Vec<String>>>,

    // Cancellation
    generation: Arc<AtomicU64>,
    cancelled: Arc<AtomicBool>,

    // State
    is_generating: Arc<AtomicBool>,
}

impl LlmClient {
    /// Create a new LLM client
    pub fn new(cfg: &LlmConfig) -> Self {
        Self {
            url: cfg.url.clone(),
            model: cfg.model.clone(),
            temperature: cfg.temperature,
            system_prompt: cfg.system_prompt.clone(),
            client: Client::new(),
            pending_sentences: Arc::new(Mutex::new(Vec::new())),
            generation: Arc::new(AtomicU64::new(0)),
            cancelled: Arc::new(AtomicBool::new(false)),
            is_generating: Arc::new(AtomicBool::new(false)),
        }
    }

    /// Generate LLM response for given messages
    ///
    /// Streams the response and splits on sentence boundaries.
    /// Complete sentences are added to pending_sentences.
    pub async fn generate(&self, messages: Vec<ChatMessage>) -> Result<()> {
        self.cancelled.store(false, Ordering::SeqCst);
        self.is_generating.store(true, Ordering::SeqCst);

        let gen_at_start = self.generation.load(Ordering::SeqCst);

        let request = ChatCompletionRequest {
            model: self.model.clone(),
            messages,
            temperature: self.temperature,
            max_tokens: -1,
            stream: true,
        };

        debug!("Starting LLM generation");

        let response = self
            .client
            .post(&self.url)
            .json(&request)
            .send()
            .await
            .context("Failed to send LLM request")?;

        if !response.status().is_success() {
            let status = response.status();
            let text = response.text().await.unwrap_or_default();
            error!("LLM API error: {} - {}", status, text);
            self.is_generating.store(false, Ordering::SeqCst);
            return Err(anyhow::anyhow!("LLM API error: {}", status));
        }

        let mut stream = response.bytes_stream();
        let mut accumulated_text = String::new();
        let mut buffer = String::new();

        while let Some(chunk_result) = stream.next().await {
            // Check cancellation
            if self.cancelled.load(Ordering::SeqCst) {
                debug!("LLM generation cancelled");
                break;
            }

            // Check generation (in case abort happened mid-stream)
            if gen_at_start != self.generation.load(Ordering::SeqCst) {
                debug!("LLM generation aborted (generation mismatch)");
                break;
            }

            let chunk = match chunk_result {
                Ok(c) => c,
                Err(e) => {
                    warn!("Stream error: {}", e);
                    continue;
                }
            };

            // Add to buffer and process complete lines
            buffer.push_str(&String::from_utf8_lossy(&chunk));

            // Process complete lines
            while let Some(newline_pos) = buffer.find('\n') {
                let line = buffer[..newline_pos].trim().to_string();
                buffer = buffer[newline_pos + 1..].to_string();

                if line.is_empty() || !line.starts_with("data: ") {
                    continue;
                }

                let data_str = &line[6..]; // Remove "data: " prefix

                if data_str == "[DONE]" {
                    debug!("LLM stream complete");
                    break;
                }

                // Parse JSON
                if let Ok(chunk) = serde_json::from_str::<StreamChunk>(data_str) {
                    if let Some(choice) = chunk.choices.first() {
                        if let Some(content) = &choice.delta.content {
                            accumulated_text.push_str(content);

                            // Check for sentence boundaries and emit complete sentences
                            self.extract_and_emit_sentences(&mut accumulated_text, gen_at_start)
                                .await;
                        }
                    }
                }
            }
        }

        // Emit any remaining text as final sentence
        if !accumulated_text.trim().is_empty() {
            if !self.cancelled.load(Ordering::SeqCst)
                && gen_at_start == self.generation.load(Ordering::SeqCst)
            {
                let mut pending = self.pending_sentences.lock().await;
                pending.push(accumulated_text.trim().to_string());
                debug!("Added final sentence: {}", accumulated_text.trim());
            }
        }

        self.is_generating.store(false, Ordering::SeqCst);
        debug!("LLM generation finished");

        Ok(())
    }

    /// Extract complete sentences and add to pending
    async fn extract_and_emit_sentences(&self, text: &mut String, gen_at_start: u64) {
        // Find sentence boundaries (. ! ?)
        let mut last_boundary = 0;

        for (i, c) in text.char_indices() {
            if c == '.' || c == '!' || c == '?' {
                // Check if this is followed by a space or end of text
                let next_idx = i + c.len_utf8();
                if next_idx >= text.len()
                    || text[next_idx..].starts_with(' ')
                    || text[next_idx..].starts_with('\n')
                {
                    // Found sentence boundary
                    let sentence = text[last_boundary..=i].trim();
                    if !sentence.is_empty() {
                        // Check generation before adding
                        if gen_at_start == self.generation.load(Ordering::SeqCst) {
                            let mut pending = self.pending_sentences.lock().await;
                            pending.push(sentence.to_string());
                            debug!("Sentence complete: {}", sentence);
                        }
                    }
                    last_boundary = next_idx;
                }
            }
        }

        // Keep only the incomplete part
        *text = text[last_boundary..].to_string();
    }

    /// Get all ready sentences (clears the list)
    pub async fn get_ready_sentences(&self) -> Vec<String> {
        let mut pending = self.pending_sentences.lock().await;
        std::mem::take(&mut *pending)
    }

    /// Check if there are pending sentences
    pub async fn has_pending_sentences(&self) -> bool {
        !self.pending_sentences.lock().await.is_empty()
    }

    /// Abort current generation
    pub fn abort(&self) {
        // Increment generation (invalidates in-flight responses)
        self.generation.fetch_add(1, Ordering::SeqCst);
        // Signal cancellation
        self.cancelled.store(true, Ordering::SeqCst);
        // Clear pending sentences synchronously is tricky, we'll do it async
        debug!("LLM abort signaled");
    }

    /// Clear pending sentences
    pub async fn clear_pending(&self) {
        self.pending_sentences.lock().await.clear();
    }

    /// Check if currently generating
    pub fn is_generating(&self) -> bool {
        self.is_generating.load(Ordering::SeqCst)
    }

    /// Check if processing (generating or has pending)
    pub async fn is_processing(&self) -> bool {
        self.is_generating() || !self.pending_sentences.lock().await.is_empty()
    }

    /// Get system prompt
    pub fn system_prompt(&self) -> &str {
        &self.system_prompt
    }
}
